File: LeetCode/力扣杯/LCP 30. 魔塔游戏.cpp
using LL = long long;
class Solution {
public:
    int magicTower(vector<int>& nums) {
        LL sum = 0;
        for(int x : nums) sum += x;
        if(sum < 0) return -1;
        sum = 0;
        priority_queue<int> pq;
        int ans = 0;
        for(int x : nums){
            if(x > 0) sum += x;
            else{
                sum += x;
                pq.push(-x);
                while(sum < 0){
                    sum += pq.top();
                    pq.pop();
                    ans += 1;
                }
            }
        }
        return ans;
    }
};

File: LeetCode/力扣杯/LCP 30. 魔塔游戏.py
class Solution:
    def magicTower(self, nums: List[int]) -> int:
        cur, monster, res = 1, 0, 0
        q = []
        for x in nums:
            cur += x
            if x < 0:
                heapq.heappush(q,x)
            while q and cur <= 0:
                m = heapq.heappop(q)
                cur -= m 
                monster += m 
                res += 1
            if cur <= 0:
                return -1
        if cur + monster <= 0:
            return -1
        return res 
